pkg/operator/management/management_state_controller.go

```diff
@@ -0,0 +1,170 @@
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+
+	configv1helpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+var workQueueKey = "instance"
+
+type ManagementStateController struct {
+	clusterOperatorName string
+	relatedObjects      []configv1.ObjectReference
+
+	operatorClient        operatorv1helpers.OperatorClient
+	clusterOperatorClient configv1client.ClusterOperatorsGetter
+	eventRecorder         events.Recorder
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewOperatorManagementStateController(
+	name string,
+	clusterOperatorClient configv1client.ClusterOperatorsGetter,
+	operatorStatusProvider operatorv1helpers.OperatorClient,
+	recorder events.Recorder,
+) *ManagementStateController {
+	c := &ManagementStateController{
+		clusterOperatorName:   name,
+		clusterOperatorClient: clusterOperatorClient,
+		operatorClient:        operatorStatusProvider,
+		eventRecorder:         recorder,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ManagementStateController-"+name),
+	}
+
+	operatorStatusProvider.Informer().AddEventHandler(c.eventHandler())
+	// TODO watch clusterOperator.status changes when it moves to openshift/api
+
+	return c
+}
+
+// sync reacts to a change in prereqs by finding information that is required to match another value in the cluster. This
+// must be information that is logically "owned" by another component.
+func (c ManagementStateController) sync() error {
+	detailedSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if apierrors.IsNotFound(err) {
+		c.eventRecorder.Warningf("StatusNotFound", "Unable to determine current operator status for %s", c.clusterOperatorName)
+		if err := c.clusterOperatorClient.ClusterOperators().Delete(c.clusterOperatorName, nil); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	originalClusterOperatorObj, err := c.clusterOperatorClient.ClusterOperators().Get(c.clusterOperatorName, metav1.GetOptions{})
+	if err != nil {
+		// Wait for the status controller to create status for us
+		glog.Infof("unable to get operator status: %v", err)
+		return err
+	}
+
+	clusterOperatorObj := originalClusterOperatorObj.DeepCopy()
+
+	clusterOperatorObj.Status = configv1.ClusterOperatorStatus{}
+
```

Comment:
```go
if IsOperatorAlwaysManaged() && detailedSpec.ManagementState == operatorv1.Unmanaged{
    setStatus(c.operatorClient, condition{"managementStateFailing", Message: "illegal state"})
}
if IsOperatorUnknownState{
    setStatus(c.operatorClient, condition{"managementStateFailing", Message: "unknown state"})
}
```

Reply:
@deads2k is it worth? it will make the code more complex (I need deep equal and unmanage is setting more conditions)